File: sort/SelectionSort/selectionSort.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char** argv){
    if (argc != 2){
        printf("Please provide file name");
        return 1;
    }

    char* fileName = argv[1];

    FILE *fp = fopen(fileName, "r");

    int numOfElems;
    fscanf(fp, "%d\n", &numOfElems);

    char* myTextArray = malloc(sizeof(char) * numOfElems);
    
    // Fill up myTextArray
    for (int i = 0; i < numOfElems; i++){
        fscanf(fp, "%c\n", &myTextArray[i]);
        // printf("%c\n", myTextArray[i]);
    }

    fclose(fp);
	printf("num: %d\n", numOfElems);

	// Open file for writing
    fp = fopen("selectionSortCOutput.txt", "w");
    char temp;
    // Selection Sort
    int smallestCharIndex; 
    for (int i = 0; i < numOfElems - 1; i++){
        smallestCharIndex = i;
        for (int j = i + 1; j < numOfElems; j++){
            if (myTextArray[smallestCharIndex] > myTextArray[j]){
                smallestCharIndex = j;
            }
        }
        // Swap smallest with current
        temp = myTextArray[i];
        myTextArray[i] = myTextArray[smallestCharIndex];
        myTextArray[smallestCharIndex] = temp;
        // myTextArray[i], myTextArray[smallestCharIndex] = myTextArray[smallestCharIndex], myTextArray[i]; // shorthand doesn't work, works in python though
        // printf("%c ", myTextArray[i]);
		// Write to file rather than print
		fprintf(fp, "%c\n", myTextArray[i]);
		//printf("%c\n", myTextArray[i]);
		if (i == numOfElems - 2){
			fprintf(fp, "%c\n", myTextArray[i + 1]);
		}
    }
	

    printf("myTextArray[0] = %c\n", myTextArray[0]);
	
	// remember to close file
    fclose(fp);

    return 0;
}

File: sort/SelectionSort/selectionSort.py
import sys # to get command line arguments


fileName = sys.argv[1]

f = open(fileName, "r")


numOfElems = 0
numOfElems = int(f.readline().strip()) # cast to int type after taking off "\n"

myTextList = [0] * numOfElems # this is a list, not an array. Think of lists as resizable arrays

# Fill up myTextList
for i in range(numOfElems):
    myTextList[i] = f.readline().strip() # strip off \n, the newline character
#print(myTextList)
f.close()

# similarly, rather than doing this, since myTextList is a list and not an array, we no longer need the number
# which tells us how many elements there are, as we do not need to preallocate memory for it.
f = open(fileName, "r")
myTextListTest = []
while True:
    element = f.readline().strip()
    # if not element: # not empty string, short hand for it
    if not element != "":
        break
    myTextListTest.append(element) 
    
print(myTextListTest) # we would expect the 10 to be part of the elements, simple if statement to ignore it,
                      # we can see from this though that we no longer need the nmber of elements.
f.close()

# Selection Sort
smallestCharIndex = 0
for i in range(numOfElems - 1):
    smallestCharIndex = i
    for j in range(i + 1, numOfElems):
        if (myTextList[j] < myTextList[smallestCharIndex]): # strictly smaller, not <= because we want to preserve order.
            smallestCharIndex = j
    
    # swap smallest with current
    myTextList[i], myTextList[smallestCharIndex] = myTextList[smallestCharIndex], myTextList[i]

print(myTextList)

f = open("selectionSortPyOutput.txt", "w")
for i in range(numOfElems):
    f.write(myTextList[i])

